Implement second half of the pipeline in pypeline

mrjackv opened the pull request.

mrjackv committed:
Add `RemoveLiftingArtifacts` to pypeline

mrjackv committed:
Add `PromoteInitCSVToUndef` to pypeline

mrjackv committed:
Add `InjectStackSizeProbesAtCallSites` to pypeline

mrjackv committed:
Add `PromoteStackPointer` to pypeline

mrjackv committed:
Add `detect-stack-size` to pypeline

mrjackv committed:
Add `SimplifySwitch` to pypeline

mrjackv committed:
Add `LegacySegregateStackAccesses` to pypeline

mrjackv committed:
Introduce `LLVMFunctionMixin`

Add a mixin class for piperuns that work on an LLVM module and need
both the model function and the `llvm::Function`. Pipe that inherit this
mixin will receive both the model and LLVM function in the
`runOnLLVMFunction` method.

mrjackv committed:
Add `MakeSegmentRef` to pypeline

mrjackv committed:
Add `make-segment-ref` to pypeline

mrjackv committed:
Add `emit-c-as-single-file` to pypeline

mrjackv committed:
Add `ParseBinaryAnalysis` to pypeline

mrjackv committed:
Add `ImportWellKnownModelsAnalysis` to pypeline

mrjackv committed:
Pipebox: reorder `link_libraries`

mrjackv committed:
Add `DetectABI` analysis to pypeline

mrjackv committed:
ObjectDependenciesTracker: also stop tracking

Call `revng::Tracking::stop` in the destructor of
`ObjectDependenciesTracker` to avoid tracking happening after the object
is destroyed.

mrjackv committed:
Add `DetectStackSize` analysis to pypeline

mrjackv committed:
Outline `mergeBackImpl` in IRHelpers

Move the body of `LLVMContainer::mergeBackImpl` in
`Support/IRHelpers.cpp` under the name `linkFunctionModules`.

mrjackv committed:
Add `AnalyzeDataLayout` analysis to pypeline

mrjackv committed:
DLA: ignore uses that are not `CallInst`s

In some cases the uses of a function are not `CallInst`s (e.g.
`ptrtoint`). Since DLA is only concerned with calls, ignore all the
other uses by using the `callers` functions.

mrjackv committed:
Add `ConvertFunctionsToCABI` analysis to pypeline

mrjackv committed:
Add `initial-auto-analysis` list to pypeline

mrjackv committed:
`revng2 project analyze`: fix analysis arguments

When setting the variable names for the analysis' configuration,
sanitize the analysis' name, as it might contain dashes and other
undesirable characters.

mrjackv committed:
`linkFunctionModule`: improve linking of globals

When linking globals that are later going to be uniqued by their id,
change their name to a temporary unique value and then restore it
post-linking, this avoids situations where function declarations shared
a common name but not metadata and were merged together by the linker.

mrjackv committed:
`makeGlobalObjectsArray`: use opaque pointers

Switch from using an array of integers with `ptrtoint` to using an array
of opaque pointers to store globals in an array to prevent their
deletion. This avoids spurious `ptrtoint`s when looking at uses of
globals.

mrjackv committed:
Drop `LLVMFunctionPassPipe`

Drop the `LLVMFunctionPassPipe` as it will not be used in the pipeline
by any pipe and its infrastructure is not needed.

mrjackv committed:
Make `SingleOutputPipe` uniform to other piperuns

Make the interface required by `SingleOutputPipe` similar to other
piperuns by requiring a constructor, the use of `PipeRunArgument`s and
the implementation of a non-static `run` function.

mrjackv committed:
Add `SegregateStackAccesses` to pypeline

mrjackv committed:
Add `SwitchToStatements` to pypeline

mrjackv committed:
Add `CliftFunctionContainer`

Add a container that owns one `mlir::ModuleOp` per function, to be used
in conjunction with the new clift backend.

mrjackv committed:
Add `LLVMToClift` to pypeline

mrjackv committed:
Introduce `CliftFunctionMixin`

In equivalence with `LLVMFunctionMixin`, add a mixin that allows
writing a pipe that instead of defining the `runOnFunction` method it
defines the `runOnCliftFunction` which receives both the
`module::Function` and the `FunctionOp`.

mrjackv committed:
Add `PureMLIRPassesPipe` to pypeline

In the same vein as `PureLLVMPassesPipe`, add a pipe to pypeline that
allows running arbitrarily-specified MLIR passes that do not depend on
the model.

mrjackv committed:
Add `ModelVerifyClift` to pypeline

mrjackv committed:
Add `ImportModelNames` to pypeline

mrjackv committed:
Add `EmitC` to pypeline

mrjackv committed:
Add `emit-c` to pypeline

mrjackv committed:
pypeline: fix `Request.check`

The function `Request.check` did not work properly due to
`Container.contains_all` performing the check with the operands swapped.
Fix the call and improve the body of `ObjectSet.issubset`.
Also improve the error message, highlighting the missing objects.

mrjackv committed:
Add `MergeLLVMModules` to pypeline

mrjackv committed:
Add `CleanupIR` to pypeline

mrjackv committed:
Add `cleanup-ir` to pypeline

mrjackv committed:
Add `emit-cfg` artifact to pypeline

mrjackv committed:
Add `ProcessCallGraph` to pypeline

mrjackv committed:
Add `YieldCallGraph` to pypeline

mrjackv committed:
Add `yield-call-graph` to pypeline

mrjackv committed:
Add `yield-call-graph-slice` to pypeline

mrjackv committed:
Add `yield-cfg` to pypeline

mrjackv committed:
pypeline: also consider `image/svg` as text

mrjackv committed:
Implement `revng2 project init`

mrjackv committed:
Implement `revng2 quick artifact`

aleclearmind commented on python/revng/pypeline/cli/project/analyze.py:
```diff
@@ -121,7 +121,7 @@ def _build_analysis_list_command(self, analysis_list_name: str, pipeline: Pipeli
             run_analysis_command = click.option(
                 f"--{normalize_flag(analysis_name)}-configuration",
-                f"{analysis_name}_configuration",
+                f"{normalize_pos_arg_name(analysis_name)}_configuration",
```
The CLI arguments use `-`, not `_`, correct?

mrjackv replied:
Yes, but the variable where this is stored needs to be snake_case

aleclearmind commented on python/revng/pypeline/container.py:
```diff
@@ -183,7 +183,7 @@ def contains_all(self, obj: ObjectSet) -> bool:
             self.kind == obj.kind
         ), f"Container {self} has kind {self.kind}, but the object set has kind {obj.kind}."
-        return self.objects().issubset(obj)
+        return obj.issubset(self.objects())
```
Ouch

aleclearmind commented on include/revng/PipeboxCommon/Helpers/PipeRunPipes/LLVMFunctionPassPipe.h:
```diff
@@ -1,229 +0,0 @@
-                                      revng::pypeline::LLVMFunctionContainer>;
-  using ContainerTypesRef = detail::TupleWithRef<typename Base::ContainerTypes>;
-
-  class Pass : public llvm::ModulePass {
```
:crying_cat_face:

aleclearmind commented on include/revng/Recompile/CompileModulePipe.h:
```diff
@@ -64,19 +64,30 @@ using ObjectFileContainer = BytesContainer<"ObjectFileContainer",
+                                             "The compiled object file",
+                                             Access::Write>>;
+
+  CompileRootModule(const Model &TheModel,
+                    llvm::StringRef StaticConfig,
+                    llvm::StringRef DynamicConfig,
+                    LLVMRootContainer &Input,
+                    ObjectFileContainer &Output);
+
+  void run();
```
The tracking starts at the constructor, right?
It's not just tracking what's done in `run`, correct?

mrjackv replied:
It's done from the constructor onward

aleclearmind commented on include/revng/ADT/STLExtras.h:
```diff
@@ -448,6 +448,9 @@ struct add_reference {
 using add_reference_t = add_reference<T>::type;
 
+template<typename T, typename V>
+using mirror_const = std::conditional_t<std::is_const_v<T>, const V, V>;
```
1. We have `ConstIf`.
2. Use snake_case only if we need to be compatible with the STL.

mrjackv replied:
1. Done
2. Done

aleclearmind commented on include/revng/Model/IRHelpers.h:
```diff
@@ -63,3 +65,30 @@ inline llvm::IntegerType *getLLVMTypeForScalar(llvm::LLVMContext &Context,
+template<typename T>
+  requires(std::same_as<T, llvm::Module> or std::same_as<T, const llvm::Module>)
+decltype(auto)
+getUniqueIsolatedFunction(T &Module, const MetaAddress *Address = nullptr) {
```
1. ```
   ConstOrNot<llvm::Module> auto &Module, const MetaAddress *Address = nullptr)
   ```
2. Do you actually use the default argument?
3. In general, I'd avoid sending in a `nullptr`, you can use `MetaAddress::invalid`. For data structures that have an invalid state, it's better to use it instead of `nullptr` or `std::nullopt`.

mrjackv replied:
1. Ok
2. Yes, in `SimplifySwitchPass`
3. Ok, even though tecnically you could have an isolated function (in the IR sense) that has an invalid MA

mrjackv replied:
1. Done
2. .
3. Switched to `MetaAddress::invalid()`

aleclearmind commented on include/revng/PipeboxCommon/Helpers/PipeRunPipes/Base.h:
```diff
@@ -58,14 +58,27 @@ concept HasPipeRunInvalidate = requires(const T &A,
 
 } // namespace detail
 
+template<typename T>
```
Concept? :thinking:

mrjackv replied:
This has been dropped

aleclearmind commented on lib/FunctionIsolation/EnforceABI.cpp:
```diff
@@ -561,11 +561,9 @@ static pipeline::RegisterPipe<EnforceABIPipe> Y;
-                                                      .llvmName(TheFunction));
-
+void EnforceABI::runOnFunction(const model::Function &Function,
+                               llvm::Function &LLVMFunction) {
```
Very nice

aleclearmind commented on include/revng/PipeboxCommon/Helpers/PipeRunPipes/Base.h:
```diff
@@ -58,14 +58,27 @@ concept HasPipeRunInvalidate = requires(const T &A,
 
 template<SingleOutputPipeBaseCompatible T>
-class SingleOutputPipeBase {
+class SingleOutputPipeBase : protected SingleOutputPipeOutputTraits<T> {
```
Why `protected`?

mrjackv replied:
Basically the part that was taken out and made into `SingleOutputPipeOutputTraits` was the `protected` part that was previously present in this class

aleclearmind commented on include/revng/PipeboxCommon/Helpers/PipeRunPipes/LLVMFunctionPipeRun.h:
```diff
@@ -0,0 +1,72 @@
+// that are exposed by the PipeRun (e.g. `Arguments`) and a few optional
+// methods that might be exposed (e.g. `checkPrerequisites`) this class
+// inherits the PipeRun so that it doesn't have to "proxy" everything, which
+// would be required if the class was encapsulated into this one.
```
1. Use Doxygen.
2. Split in multiple paragraphs.

mrjackv replied:
This has been dropped

aleclearmind commented on include/revng/PipeboxCommon/Helpers/PipeRunPipes/Base.h:
```diff
@@ -58,14 +58,27 @@ concept HasPipeRunInvalidate = requires(const T &A,
 } // namespace detail
 
+template<typename T>
+struct SingleOutputPipeOutputTraits {
```
1. I don't understand what this is. Sometimes you inherit from it (weird inheriting from a trait, unless you're a trait as well), sometimes you just use it.
   Let's discuss this in person maybe.

aleclearmind commented on include/revng/PipeboxCommon/Helpers/PipeRunPipes/LLVMFunctionPipeRun.h:
```diff
@@ -0,0 +1,72 @@
+// inherits the PipeRun so that it doesn't have to "proxy" everything, which
+// would be required if the class was encapsulated into this one.
+template<IsLLVMFunctionPipeRun T>
+class LLVMFunctionPipeRun : public T {
```
1. I'm not sure what this is. Does it respect `IsFunctionPipe`?
   If so, `static_assert` in its body.
2. (unrelated) The `IsFunctionPipe` concept should be called `IsFunctionPipeRun`, correct?
   Please review all concept names in this sense. Also, all those that respect this (and similar) concepts should be either have a `*PipeRun` suffix or (better) be in a namespace that makes it obvious that these are pipe runs.
3. `T` is not a pipe run strictly speaking (I don't think it respects the `IsFunctionPipe` concept). Add a note above the `IsLLVMFunctionPipeRun` concept about this. Same for the MLIR one.

mrjackv replied:
I've rotated the inheritance, now there's `LLVMFunctionMixin<Derived>` and `MLIRFunctionMixin<Derived>`

aleclearmind commented on lib/RemoveLiftingArtifacts/RemoveLiftingArtifacts.cpp:
```diff
@@ -225,3 +227,14 @@ struct RemoveLiftingArtifactsPipe {
+
+namespace revng::pypeline::piperuns {
+
+void RemoveLiftingArtifacts::runOnFunction(const model::Function &Function,
```
Please review all the pipe runs you implemented and add something like

```cpp
// TODO: merge ::RemoveLiftingArtifacts into RemoveLiftingArtifacts once we dismiss the old pipeline
```

Do this for all the situations where this is obvious.

mrjackv replied:
Done

aleclearmind commented on include/revng/RemoveLiftingArtifacts/PromoteInitCSVToUndef.h:
```diff
@@ -0,0 +1,33 @@
+  static constexpr llvm::StringRef Name = "promote-init-csv-to-undef";
+  using Arguments = TypeList<PipeRunArgument<LLVMFunctionContainer,
+                                             "Module",
+                                             "function LLVM module(s)">>;
```
I think we should hoist in some header the `PipeRunArgument` instantiations that are identical all over the place.

mrjackv replied:
Done

aleclearmind commented on lib/PromoteStackPointer/InjectStackSizeProbesAtCallSitesPass.cpp:
```diff
@@ -65,3 +71,15 @@ char InjectStackSizeProbesAtCallSitesPass::ID = 0;
+void InjectStackSizeProbesAtCallSites::runOnFunction(const model::Function
+                                                       &Function) {
+  llvm::Module &Module = ModuleContainer.getModule(ObjectID(Function.Entry()));
+  GeneratedCodeBasicInfo GCBI(Binary);
+  GCBI.run(Module);
```
Ah-ah, so this is the trick.

aleclearmind commented on lib/Canonicalize/SimplifySwitchPass.cpp:
```diff
@@ -279,3 +281,60 @@ void SimplifySwitch::run(pipeline::ExecutionContext &EC,
+  auto BinaryBuffer = BinariesContainer.getFile(0);
+  BinaryView.emplace(Binary,
+                     llvm::StringRef{ BinaryBuffer.data(),
+                                      BinaryBuffer.size() });
```
Illegal usage of `std::optional`! Use an helper function and do `BinaryView(myHelperFunction(...))`.

mrjackv replied:
Done

aleclearmind commented on lib/Canonicalize/SimplifySwitchPass.cpp:
```diff
@@ -279,3 +281,60 @@ void SimplifySwitch::run(pipeline::ExecutionContext &EC,
+
+  bool runOnModule(llvm::Module &M) override {
+    llvm::Function &Function = getUniqueIsolatedFunction(M);
+    auto &LVI = getAnalysis<LazyValueInfoWrapperPass>(Function).getLVI();
```
So, in the end you didn't manage to instantiate LVI without going through the analysis route?
For the `DominatorTree` it's quite straightforward.

At very least explain why in a comment and/or leave a TODO.

mrjackv replied:
> So, in the end you didn't manage to instantiate LVI without going through the analysis route?

There are a few dependencies and they nest quite deeply, at some point I thought something along the lines of "Ah I'll write something that automatically computes the dependencies for me" and that's where I've said "why should I do all of this when there `PassManager` already?"

> At very least explain why in a comment and/or leave a TODO

Will do

aleclearmind commented on python/revng/internal/pipeline.yml:
```diff
@@ -198,6 +198,34 @@ branches:
+      - pipe: pure-llvm-passes-pipe
+        arguments: [llvm-functions]
+        configuration:
+          Passes:
```
Can we have a field named `Passes` in C++ but the YAML serializer use `passes`?

mrjackv replied:
Done

aleclearmind commented on include/revng/RemoveLiftingArtifacts/MakeSegmentRef.h:
```diff
@@ -0,0 +1,45 @@
+                 LLVMFunctionContainer &ModuleContainer) :
+    Binary(*Model.get().get()) {
+    llvm::ArrayRef<char> Buffer = BinariesContainer.getFile(0);
+    BinaryView.emplace(Binary, llvm::StringRef{ Buffer.data(), Buffer.size() });
```
Illegal use of `std::optional`, please fix everywhere.

mrjackv replied:
Fixed and checked

aleclearmind commented on lib/RemoveLiftingArtifacts/MakeSegmentRefPass.cpp:
```diff
@@ -32,13 +32,21 @@
 
 using namespace llvm;
 
+static cl::opt<bool> DisableFunctionPointersOpt("disable-function-pointers",
+                                                cl::desc("Disable emitting "
+                                                         "pointers to "
+                                                         "functions through "
+                                                         "segments"),
+                                                cl::init(false));
```
Add a comment explaining why we do this and note that once we rewrite the `emit-segment-refs` this can go.

mrjackv replied:
Added

aleclearmind reviewed (commented):
NOTETOSELF: resume reviewing starting from `Add ParseBinaryAnalysis to pypeline`, included.

aleclearmind commented on include/revng/PipeboxCommon/Helpers/Registrars.h:
```diff
@@ -194,4 +195,7 @@ template<typename T>
 
 template<typename T>
-using RegisterLLVMFunctionPassPipe = RegisterPipe<LLVMFunctionPassPipe<T>>;
+using RegisterLLVMFunctionPipe = RegisterFunctionPipe<LLVMFunctionPipeRun<T>>;
```
`RegisterFunctionPipeRun`

mrjackv replied:
Done

aleclearmind commented on include/revng/PipeboxCommon/Helpers/Registrars.h:
Concept

mrjackv replied:
Added them

aleclearmind commented on include/revng/PipeboxCommon/Helpers/PipeRunPipes/LLVMFunctionPipeRun.h:
```diff
@@ -0,0 +1,72 @@
+// inherits the PipeRun so that it doesn't have to "proxy" everything, which
+// would be required if the class was encapsulated into this one.
+template<IsLLVMFunctionPipeRun T>
+class LLVMFunctionPipeRun : public T {
```
Rotate inheritance. Turn `LLVMFunctionPipeRun` into a mixin/CRTP so that what's currently `T` becomes actually a pipe run and we no longer have a "semi-pipe run".

mrjackv replied:
Done

mrjackv committed:
Registrars: add `Run` suffix for piperuns

Change the `Register.*Pipe` functions to end in `Run` if the template
argument should be a pipe run. Also change the concepts `Is.*Pipe` to
also have the `Run` suffix if applicable.

mrjackv committed:
pypeline-comparison: omit metadata in `lift`

Do not compare metadata when comparing `lift` as the output is still
non-deterministic in some instances.

mrjackv committed:
LLVMPipe: make passes key lowercase

Since the rest of the pypeline yaml file is lowercase, convert the
`Passes` key of the configuration used by `pure-llvm-passes-*` to
lowercase.

mrjackv committed:
Comment `run*` function with inlinable content

Add `TODO` comments to piperuns' `run*` functions that will have content
inlined once the old pipeline is removed.

aleclearmind commented on lib/Canonicalize/SimplifySwitchPass.cpp:
```diff
@@ -279,3 +282,60 @@ void SimplifySwitch::run(pipeline::ExecutionContext &EC,
+};
+
+using SSP = SimplifySwitchPass;
+static inline llvm::RegisterPass<SSP> SSPR{ "pypeline-simplify-switch",
```
I *think* you can avoid registering this if you don't want to expose it.

mrjackv replied:
Dropped

aleclearmind commented on lib/Canonicalize/SimplifySwitchPass.cpp:
```diff
@@ -279,3 +282,60 @@ void SimplifySwitch::run(pipeline::ExecutionContext &EC,
+  ModuleContainer(ModuleContainer),
+  Binary(*Model.get().get()),
+  BinaryView(makeBinaryView(Model, BinariesContainer)) {
+  PM.add(new DominatorTreeWrapperPass);
```
1. Do you *have to* schedule `DomintatorTree`? You're not scheduling `LazyValueInfo`.
2. Instantiate `DominatorTree` manually, do not depend on the analysis. Let's try to keep the number of analyses to a minimum, so that if we decide to drop LVI, we can easily demote this to a non-`Pass`.

mrjackv replied:
Instantiated the `DominatorTree` manually

aleclearmind commented on include/revng/Canonicalize/SimplifySwitch.h:
```diff
@@ -0,0 +1,50 @@
+                    "Module",
+                    "function LLVM module(s)">>;
+  using Analyses = TypeList<llvm::LazyValueInfoWrapperPass,
+                            llvm::DominatorTreeWrapperPass>;
```
Drop?

mrjackv replied:
It's already dropped, will close this in the next push

aleclearmind commented on include/revng/ADT/CompilationTime.h:
```diff
@@ -200,4 +200,34 @@ template<auto &V>
+template<typename CallableT, typename... BoolType>
+  requires(std::is_same_v<BoolType, bool> and ...)
+inline constexpr auto
+bools(CallableT &&Callable, bool Bool, BoolType... BoolRest) {
```
1. You are an evil person. But I like the way you think.
2. I'd call this `invokeAllCombinations` (or is it *permutations*?).

mrjackv replied:
Renamed to `invokeCombination`

aleclearmind commented on include/revng/Model/Importer/Binary/BinaryImporter.h:
```diff
@@ -16,9 +16,16 @@ class ObjectFile;
+importBinary(TupleTree<model::Binary> &Model,
+             llvm::object::ObjectFile &BinaryHandle,
+             const ImporterOptions &Options,
+             TupleTreeReference<model::BinaryIdentifier, model::Binary>
```
Use `BinaryReference` everywhere you can.

mrjackv replied:
That wold require having a header with `using BinaryReference = TupleTreeReference<model::BinaryIdentifier, model::Binary>`. I was under the impression that `using`s in header were prohibited, did that change?

aleclearmind replied:
`using`s are not prohibited in headers. `using namespace` are prohibited.

mrjackv replied:
Done

aleclearmind commented on include/revng/Model/Importer/Binary/ImportBinaryAnalysis.h:
```diff
@@ -29,3 +31,17 @@ class ImportBinaryAnalysis {
+
+namespace revng::pypeline::analyses {
+
+class ParseBinaryAnalysis {
```
Docs

mrjackv replied:
Done

aleclearmind commented on lib/Model/Importer/Binary/ELFImporter.cpp:
```diff
@@ -420,7 +420,15 @@ Error ELFImporter<T, HasAddend>::import(const ImporterOptions &Options) {
-    Importer.import(TheBinary.getFileName(), AdjustedOptions);
+    auto Buffer = MemoryBuffer::getMemBuffer(TheBinary.getData(), "", false);
+
+    std::string Filename;
+    if (Reference.isValid())
+      Filename = Reference.get()->Name();
+    else if (not TheBinary.getFileName().empty())
+      Filename = TheBinary.getFileName().str();
```
What is going on here?

mrjackv replied:
`ELFImporter` previously relied on the `TheBinary` field to retrieve the filename of the file. This is no longer possible because the API has been changed to work with `llvm::MemoryBuffer` as it's friendlier with `BinariesContainer`. The importer still needs the filename because it does a bunch of heuristics (e.g. looking for `${FILENAME}.debug`) to retrieve the debug info. So first the name is retrieved from `/Binaries/0/Name` if it's valid or the `TheBinary` as a fallback (which is compatible to what the old pipeline does)

aleclearmind replied:
Add a comment clearly explaining what needs to happen once we dismiss the old pipeline.

mrjackv replied:
Done

aleclearmind commented on lib/Model/Importer/WellKnownModels.cpp:
```diff
@@ -48,66 +49,85 @@ class ImportWellKnownModelsAnalysis {
+
+} // namespace revng::pipes
+
+static void wellKnownModel(TupleTree<model::Binary> &Model) {
```
This name means nothing to me

mrjackv replied:
It's the body of `WellKnownModel::run` from the legacy pipe converted to a static function so that it can be used by both the old and new pipe. I've followed the pattern of converting the `run` into a static function with a `camelCase` name consistently in the PR

aleclearmind replied:
What does it do? Functions should be a verb or a noun representing what they return.

mrjackv replied:
It does the `ImportWellKnownModels` analysis but without the `revng::pipeline`-isms

aleclearmind replied:
`importWellKnownModels`?

mrjackv replied:
renamed

aleclearmind commented on include/revng/EarlyFunctionAnalysis/CollectFunctionsFromCalleesPass.h:
```diff
@@ -12,6 +12,14 @@
 #include "revng/Support/IRHelpers.h"
 #include "revng/Support/MetaAddress.h"
 
+namespace revng::internal {
```
I don't like `revng::internal`. Why do you need it?

I feel like you're trying to circumvent some rule on the `details` namespace `:P`

mrjackv replied:
The problem is as follows:
1. There's a function or a chunk of it that needs to be exposed between TUs, this is because (as in this case, there are variations) the code was inside an LLVM pass or some function that accepted pipeline (the old one)-specific inputs
2. I outline the chunk I need in a function that's called by both the old and new pipeline stuff
3. I put it in the `revng::internal` namespace because it's a free-floating chunk of code stitched to other parts of pipes that doesn't do meaningful things on it own and I can't use `detail` since you said that it cannot be used across TUs

aleclearmind replied:
3. Please do not try to circumvent rules. Ask out. `revng::internal` is not a convention we adopt and it's just a way to circumvent a rule. These are tricks to avoid rules that risk going unnoticed during code review and then become a precedent that someone else can adopt. Refrain from doing this.

A free function is fine. If you don't like it, you can make a static method in the class.

mrjackv replied:
Moved the two functions outside the `revng::internal` namespace

aleclearmind commented on include/revng/DataLayoutAnalysis/DLA.h:
```diff
@@ -0,0 +1,22 @@
+
+namespace revng::pypeline::analyses {
+
+class DLA {
```
Do not use abbreviations. `AnalyzeDataLayout`.

mrjackv replied:
Renamed

aleclearmind commented on lib/DataLayoutAnalysis/Frontend/DLACreateInterProceduralTypes.cpp:
```diff
@@ -276,7 +276,10 @@ bool TSBuilder::createInterproceduralTypes(llvm::Module &M,
-      auto *Call = cast<CallInst>(U.getUser());
+      auto *Call = dyn_cast<CallInst>(U.getUser());
+      if (Call == nullptr)
+        continue;
```
Nope, use the `callers` function.

mrjackv replied:
Done

aleclearmind commented on include/revng/PipeboxCommon/CliftContainer.h:
@vasama-work can you review this please?

mrjackv replied:
Closing this since there are comments to address

aleclearmind commented on include/revng/Clift/Helpers.h:
```diff
@@ -22,4 +22,32 @@ inline MetaAddress getMetaAddress(clift::FunctionOp F) {
   return MetaAddress::invalid();
 }
 
+inline decltype(auto)
```
Mmmh, please be explicit (you can use `ConstIf`).

mrjackv replied:
Done

aleclearmind commented on python/revng/internal/pipeline.yml:
No abbreviations in commit messages. What's `s.p.`?

mrjackv replied:
**S**ave**p**oint
If I write it out it exceeds the 52 line limit imposed by check-conventions

aleclearmind replied:
> Add `segregate-stack-accesses` to pypeline
>
> This commit adds...

However, I doubt the usefulness of all these commits, as mentioned elsewhere.

mrjackv replied:
Compacted changes to `pipeline.yml` into a single commit for each branch of the pipeline

aleclearmind commented on lib/CliftPipes/ClifterPipe.cpp:
```diff
@@ -57,3 +59,36 @@ class LLVMToCliftPipe {
+  Output.getContext().loadDialect<clift::CliftDialect>();
+}
+
+void LLVMToClift::runOnFunction(const model::Function &Function) {
```
@vasama-work can you review this?

mrjackv replied:
Closing this since there are comments to address

aleclearmind commented on lib/CliftPipes/OptimizationPipe.cpp:
```diff
@@ -68,3 +69,25 @@ class OptimizationPipe {
+  PM(&ModuleContainer.getContext(), clift::FunctionOp::getOperationName()) {
+  PM.addPass(mlir::createCanonicalizerPass());
+  PM.addPass(clift::createLoopDetectionPass());
+  PM.addPass(clift::createCLegalizationPass(TargetCImplementation::Default));
```
I don't understand the purpose of this pipe run.
Aren't these pure MLIR passes? We should have something to run pure MLIR passes, or did you want to postpone this?

mrjackv replied:
Added `pure-mlir-passes-pipe`

aleclearmind commented on lib/CliftPipes/ModelVerifyPipe.cpp:
```diff
@@ -39,3 +40,19 @@ class ModelVerifyPipe {
+  if (not ModelVerifyLogger.isEnabled())
+    return;
+
+  revng::Tracking::push(Binary);
```
1. I think we have a RAII object for this.
2. Comment why we are doing this.

mrjackv replied:
1. `ObjectDependenciesHelper::Committer` exists but it's a private class and I don't think it's warranted to expose it since it's tightly couple with `ObjectDependenciesHelper`
2. Ok

aleclearmind replied:
1. `struct DisableTracking`. Adopt it everywhere it makes sense.

mrjackv replied:
Done

aleclearmind commented on python/revng/internal/pipeline.yml:
```diff
@@ -91,6 +91,9 @@ branches:
         arguments: [llvm-root, cfg-map]
       - savepoint: cfg-computed
         containers: [cfg-map]
+        artifacts:
+          - name: emit-cfg
+            container: cfg-map
```
I'm not entirely sure what's the value of having all these micro commits on `python/revng/internal/pipeline.yml`.

Can we have a single commit at the end?

mrjackv replied:
Fixed, see [here](https://github.com/revng/revng/pull/550#discussion_r2607086212)

aleclearmind commented on python/revng/internal/cli/revng2.py:
```diff
@@ -26,9 +41,25 @@ def quick():
+    model_name = get_singleton(Model).model_name()  # type: ignore[type-abstract]
+    model_file = Path.cwd() / model_name
+    if model_file.exists():
+        raise click.UsageError(f"File {model_name} is present, cannot run init")
```
> is already present in the current directory. Refusing to overwrite it. ~~cannot run init~~

mrjackv replied:
Done

aleclearmind commented on share/revng/test/configuration/revng/pypeline-comparison.yml:
```diff
@@ -10,7 +10,13 @@ commands:
+      pushd "$OUTPUT" &> /dev/null;
+      cp "$INPUT" input;
+      revng2 project init input;
+      mv model.yml model_init.yml;
+      revng analyze revng-initial-auto-analysis --model=model_init.yml input > model.yml;
+      rm model_init.yml;
+      popd &> /dev/null;
```
Comment

mrjackv replied:
This got cut down drastically, please check if comments are still needed

aleclearmind commented on python/revng/internal/cli/revng2.py:
```diff
@@ -26,9 +41,25 @@ def quick():
+        return
+
+    with open(model_file, "w") as f:
+        yaml.safe_dump(generate_model_with_binaries([binary]), f)
```
`revng2 project init` should run `initial-auto-analyses`.

mrjackv replied:
That's not what's specified in [REP21](https://pad.rev.ng/o37XT3QfScKKV8Wh7h6ZCw#revng-project-init)
Making it do the initial auto analysis is not a problem, but I'd prefer if there was a `--no-initial-auto-analysis` or similar to suppress it

aleclearmind replied:
> That's not what's specified in [REP21](https://pad.rev.ng/o37XT3QfScKKV8Wh7h6ZCw#revng-project-init)

Right, but I think we already said that it should do it.

> I'd prefer if there was a --no-initial-auto-analysis or similar to suppress it

Sure.

mrjackv replied:
Implemented

aleclearmind commented on python/revng/pypeline/storage/local_provider.py:
```diff
@@ -598,3 +599,44 @@ def get_custom_invalidation_data(
+        model_path = self.dir_path / model_name
+        model_path.touch()
+
+        super().__init__(self.dir_path / "db", model_path)
```
Please keep this as the very first line of the function.

mrjackv replied:
Closing, this class has been dropped

aleclearmind commented on python/revng/pypeline/storage/local_provider.py:
```diff
@@ -598,3 +599,44 @@ def get_custom_invalidation_data(
+
+        return result
+
+    def get_files_from_storage(self, requests: list[FileRequest]) -> dict[str, bytes]:
```
Are you storing the whole file in memory? Please don't, this is not a debugging tool, this will run on real stuff.

mrjackv replied:
> Are you storing the whole file in memory?

No

The `FileRequest` is composed of two fields `content` and `path`, only one of them can be non-None. In the case of CLI stuff the `path` field will be non-none. The only case where the `content` field is non-none is if the storage provider is used in conjunction with the daemon and a request to add a file arrives via HTTP (which, at that point, does not have a file associated). The content is stored as-is because:
1. It's already in RAM, we pay 0 cost in keeping it there
2. It's unlikely that this storage provider will be used with the daemon

aleclearmind replied:
Why do we have to override `put_files_in_storage` at all?
I'd really like this to just be a `LocalStorageProvider` putting things in a temporary directory as much as we can.

I think we should be able to do this without creating a new class. We should just feed the right `__init__` arguments.
Maybe we should have a `temporary://` factory that does this (and uses the vanilla `LocalStorageProvider`).

mrjackv replied:
Closing, this class has been dropped

vasama-work commented on include/revng/PipeboxCommon/CliftContainer.h:
```diff
@@ -0,0 +1,96 @@
+
+private:
+  mlir::MLIRContext Context;
+  std::map<ObjectID, mlir::OwningOpRef<mlir::ModuleOp>> Modules;
```
Why do we have multiple *modules*? Functions I would understand.

mrjackv replied:
This mirrors the behavior of `LLVMFunctionContainer`. Each isolated function is in its own module, so that there isn't the need to run `cloneFiltered` or `mergeBackImpl` when serializing/deserializing.

mrjackv replied:
Closing this as it has been discussed on mattermost

vasama-work commented on include/revng/PipeboxCommon/MLIRContainer.h:
```diff
@@ -0,0 +1,96 @@
+
+}
+
+class MLIRFunctionContainer {
```
We renamed what was originally `MLIRContainer` to `CliftContainer` because it was so Clift-specific. Is this container intended to be more general than that?

mrjackv replied:
Renamed to `CliftFunctionContainer`

vasama-work commented on include/revng/PipeboxCommon/MLIRContainer.h:
```diff
@@ -0,0 +1,96 @@
+      mlir::LogicalResult ModuleResult = Module.get().verify();
+      Result = Result & ModuleResult.succeeded();
+    }
+    return true;
```
`Result` is unused. I suspect you meant to return it.

Also consider if `Result &=` instead of `Result = Result &` is an improvement.

mrjackv replied:
Good catch! thanks!

vasama-work commented on include/revng/PipeboxCommon/MLIRContainer.h:
```diff
@@ -0,0 +1,96 @@
+  mlir::DialectRegistry Registry;
+  Registry.insert<mlir::clift::CliftDialect>();
+  return Registry;
+})();
```
If this is only used by `MLIRFunctionContainer`, why not make it a static member variable?

mrjackv replied:
Converted to static member

vasama-work commented on include/revng/CliftPipes/ClifterPipe.h:
```diff
@@ -0,0 +1,43 @@
+  const model::Binary &Binary;
+  const LLVMFunctionContainer &Input;
+  MLIRFunctionContainer &Output;
+  model::CNameBuilder NameBuilder;
```
As far as I can tell, this is never used.

mrjackv replied:
Sorry I just copy-pasted from the old pipeline, it's indeed unused. I've dropped it.

vasama-work commented on lib/CliftPipes/ClifterPipe.cpp:
```diff
@@ -57,3 +59,36 @@ class LLVMToCliftPipe {
+  const llvm::Function
+    &LLVMFunction = getUniqueIsolatedFunction(Module, Function.Entry());
+
+  using namespace mlir;
```
Not a big issue, but for consistency with the rest of the Clift code, I would prefer not applying a using directive to namespace `mlir`. I've commonly done this though:
```cpp
namespace clift = mlir::clift;
using namespace clift;
```

mrjackv replied:
Thanks for the heads up, in the end I've dropped the use of `using namespace mlir;` in the PR

aleclearmind commented on lib/Model/Importer/Binary/ELFImporter.h:
```diff
@@ -64,6 +65,7 @@ class ELFImporter : public BinaryImporterHelper, public ELFImporterBase {
   RawBinaryView File;
   TupleTree<model::Binary> &Model;
   const llvm::object::ELFObjectFileBase &TheBinary;
+  BinaryReference &Reference;
```
`Reference` is an empty term, do not use it.

mrjackv replied:
Renamed to `BinaryReference`

aleclearmind commented on lib/Model/Importer/Binary/ImportBinaryAnalysis.cpp:
```diff
@@ -16,6 +16,17 @@
+using BinaryReference = TupleTreeReference<model::BinaryIdentifier,
+                                           model::Binary>;
+
+static BinaryReference makeReference(model::Binary *Binary, size_t Index) {
```
Use `model::Binary &` unless you want to store it (e.g., in a `map`) or you actually think having it `nullptr` is useful.

mrjackv replied:
Done

aleclearmind commented on python/revng/pypeline/storage/local_provider.py:
```diff
@@ -598,3 +599,44 @@ def get_custom_invalidation_data(
+    all the data will be lost once this provider is deleted"""
+
+    def __init__(self):
+        self.directory = TemporaryDirectory()
```
Describe in a comment the layout of this directory

mrjackv replied:
Closing, this class has been dropped

aleclearmind commented on python/revng/pypeline/storage/local_provider.py:
```diff
@@ -598,3 +599,44 @@ def get_custom_invalidation_data(
         return result
+
+
+class EphemeralLocalStorageProvider(LocalStorageProvider):
```
Do we have an URL to instantiate this?

mrjackv replied:
Yes, `temporary://`

mrjackv committed:
pypeline-cli: fix autocomplete with pipebox init

Make it so the pipebox is initialized early if the pypeline CLI is in
autocomplete mode, otherwise some pipes might break.

mrjackv committed:
InitRevng: fix memory leak due to stack trace

When initializing `revng::InitRevng` it is assumed by its parent,
`InitLLVM`, that the lifetime of `Argv` and be equivalent to
static, because it might be later retrieved to print it in case of a
crash. This works with the traditional `main`, however in the pipebox
the string passed are not guaranteed to have that lifetime. This commits
moves ownership of the strings to a class that stores them to avoid this
problem.

mrjackv committed:
Adapt mass-testing infrastructure to pypeline

Make the required changes so that `revng2` can be used with mass-testing
and the per-component stats are correct.

mrjackv committed:
mass-testing: exclude paths from flamegraph

Add a new option to the mass-testing `meta.yml` that allows skipping
some paths when generating the flamegraph. In the case of revng this
allows skipping the intermediate python frames.

mrjackv committed:
mass-testing-report: names for binaries

Instead of having all the binaries in the mass-testing-report be
downloaded with the name `input`, have them use a meaningful name
instead.

aleclearmind commented on lib/PipeboxCommon/Python/_pipebox.cpp:
```diff
@@ -33,6 +33,37 @@ static void handleSignal(int SigNo) {
+    Argc = ArgvRaw.size();
+    Argv.resize(ArgvRaw.size());
+    for (size_t I = 0; I < ArgvRaw.size(); I++)
+      Argv[I] = ArgvRaw[I].c_str();
```
Any reason not to do it in `InitRevng` itself?
I'd prefer not to have this.

mrjackv replied:
> Any reason not to do it in InitRevng itself?

Because all other uses of `InitRevng` pass the argv directly from `main`, so they don't need to do this
Also, `InitRevng` is a child of `InitLLVM` so having _it_ store the argv would require breaking the child-parent relationship

aleclearmind replied:
> would require breaking the child-parent relationship

I don't see that.

```patch
diff --git a/include/revng/Support/InitRevng.h b/include/revng/Support/InitRevng.h
index 695f55e1a..271de6c05 100644
--- a/include/revng/Support/InitRevng.h
+++ b/include/revng/Support/InitRevng.h
@@ -22,6 +22,8 @@ namespace revng {
 class InitRevng : public llvm::InitLLVM {
 private:
   static inline bool Initialized = false;
+  std::vector<std::string> Args;
+  std::vector<const char *> Argv;
 
 public:
   InitRevng(int &Argc, auto **&Argv, const char *Overview) :
@@ -31,7 +33,9 @@ public:
             auto **&Argv,
             const char *Overview,
             llvm::ArrayRef<const llvm::cl::OptionCategory *> CategoriesToHide) :
-    InitLLVM(Argc, Argv, true) {
+    Args(copy(Argv, Argc)),
+    Argv(something(Args)),
+    InitLLVM(Argc, this->Argv, true) {
 
     revng_assert(not Initialized);
     Initialized = true;
```

Alternatively, you can just make `const char *Argv[Argc]` a global variable.
I know you don't like them, but I think this is fine. Initialization is supposed to happen only once.

I don't think we need a "PythonInitRevng", there's nothing python about it. Either the existing one should not assume argv have infinite life time or we should just ensure lifetime is infinite in all users.

mrjackv replied:
> I don't see that.

> ```diff
> -    InitLLVM(Argc, Argv, true) {
>+    Args(copy(Argv, Argc)),
>+    Argv(something(Args)),
>+    InitLLVM(Argc, this->Argv, true) {
>```

That cannot be done, base class constructors need to be initialized before member variables. The only way out is to make `InitRevng` look more like `PythonInitRevng` with a `std::unique_ptr<llvm::initLLVM>` or similar

> Alternatively, you can just make const char *Argv[Argc] a global variable.

This is also impractical, it would require changing all (valid) invocations of `InitRevng`. Also, the length is not known, so you'd need to have something like `static std::vector<std::string> Args`, set it and then re-wrangle it to make it work with `InitLLVM` because its constructor needs to be the first thing that's called. Again, might as well use a `std::unique_ptr<llvm::InitLLVM>`

aleclearmind replied:
> That cannot be done

Right, dunno, we can use composition instead of inheritance. There's no real reason to inherit.

> This is also impractical

Uh? I just meant

```patch
diff --git a/lib/PipeboxCommon/Python/_pipebox.cpp b/lib/PipeboxCommon/Python/_pipebox.cpp
index 37b2296af..851da4bb8 100644
--- a/lib/PipeboxCommon/Python/_pipebox.cpp
+++ b/lib/PipeboxCommon/Python/_pipebox.cpp
@@ -52,13 +52,15 @@ NB_MODULE(_pipebox, m) {
       }
     }
 
-    int Argc = ArgVector.size() + 1;
-    const char *Argv[Argc];
+    static std::vector<std::string> ArgVector2 = ArgVector;
+    int Argc = ArgVector2.size() + 1;
+    static std::vector<const char *> Argv;
+    Argv.resize(Argc);
     Argv[0] = "";
-    for (size_t I = 0; I < ArgVector.size(); I++)
-      Argv[I + 1] = ArgVector[I].c_str();
+    for (size_t I = 0; I < ArgVector2.size(); I++)
+      Argv[I + 1] = ArgVector2[I].c_str();
 
-    const char **ArgvPtr = Argv;
+    const char **ArgvPtr = &Argv[0];
     // use a capsule to call the destructor when the Python module is unloaded
     m.attr("__init_revng__") = makeCapsule<revng::InitRevng>(Argc, ArgvPtr, "");
 ```

If you really wanna keep the `PythonInitRevng` class, you should call it `ArgvOwningInitRevng`, which is quite sad.

mrjackv replied:
Ah ok, I was a bit confused. I've dropped the class and made the variables in `Initialize` static
